v2/call.js

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var read = require('../lib/read');
+var write = require('../lib/write');
+var Checksum = require('./checksum');
+var header = require('./header');
+
+module.exports.Request = CallRequest;
+module.exports.Response = CallResponse;
+
+var emptyBuffer = new Buffer(0);
+var emptyTracing = new Buffer(25);
+emptyTracing.fill(0);
+
+/* jshint maxparams:9 */
+
+// flags:1 ttl:4 tracing:24 traceflags:1 service~2 nh:1 (hk~1 hv~1){nh} csumtype:1 (csum:4){0,1} arg1~2 arg2~2 arg3~2
+function CallRequest(flags, ttl, tracing, service, headers, csum, arg1, arg2, arg3) {
+    if (!(this instanceof CallRequest)) {
+        return new CallRequest(flags, ttl, tracing, service, headers, csum, arg1, arg2, arg3);
+    }
+    var self = this;
+    // TODO Checksum-like class for tracing
+    self.type = CallRequest.TypeCode;
+    self.flags = flags || 0;
+    self.ttl = ttl;
+    self.tracing = tracing || emptyTracing;
+    self.service = service || emptyBuffer;
+    self.headers = headers;
+    self.csum = Checksum.objOrType(csum);
```

Comment:
no callers do not get to specify the checksum themselves, merely the type. the design here is that the actual checksum value gets computed JIT during write.